### ISSUE TITLE
Debounce (protect against accidental multiple operations)

### ISSUE THREAD

**Issue by subzero911** [marcglasberg/async_redux#75]
How to implement protection against too often heavy operations?
For instance, if I pressed the "Save" button quickly several times. I don't want to save multiple times.
I want to wait a 500 ms after last press and save only once.

In MobX we have a ready-made reaction 'delay' parameter (check the 101-105 lines):
https://github.com/brianegan/flutter_architecture_samples/blob/master/mobx/lib/stores/todo_store.dart

In a pure Provider architecture I'd use an RxDart debounce() to control saving:
```dart
class Store with ChangeNotifier {
final _saveController = StreamController<item>(); //helper stream
final List<Item> _list = []; //data

Store() { _saveController.stream.debounce(ms: 500).listen => // do your save here  }

 void add(item) { //action
 _list.add(item);
 notifyListeners();
  _saveController.add(item); //add event
 }
}
```

But I have no idea, how to create similar functionality with Async Redux, because it's not recommended to keep streams in the store.

**Comment by marcglasberg** [comment]
Please, first read this section of the documentation: https://pub.dev/packages/async_redux#progress-indicators 

There are many ways to do that:

1)  You can use the `Wait` class in the store to create a modal barrier (in the UI) to prevent the button from being tapped, or to turn off the button's `onTap` callback, while the save is being performed.

2) Actions have an `abortDispatch` method. You can use the `Wait` class in the store to abort the dispatch while the save is being performed (or just return `null` from the reducer).

3) You can add a `static bool isSaving` field to the action. Make it `true` when the action starts (using the action's `before` method), and make it `false` when it finishes (using the action's `after` method). Then abort the dispatch if the bool is already `true`. Note the `abortDispatch` method runs before the `before` method.

4) If you just want to have a time-based debouncing that cancels actions before some elapsed time, add a static `static DateTime last` field to the action. Save the action's `stateTimestamp` to this variable when the action starts (using the action's `before` method), and make it `null` when it finishes (using the action's `after` method). Then abort the dispatch if the action's `stateTimestamp` is not at least `x` seconds from saved timestamp.

5) However, if you want to have a time-based debouncing that groups actions within some elapsed time, it's a bit more complex. The store itself does this when it needs to call the state persistor. But it's also more complex because it checks not only the elapsed time, but also if the previous save has finished. I guess I could add some more functionality to the AsyncRedux's `Wait` class to allow for these more complex deboucings, but I'm not sure it's necessary. Usually I just prevent the button from being tapped during the save process, in the UI, using the `Wait` class, and that's usually better than a time-based debouncing.

**Comment by subzero911** [comment]
> Usually I just prevent the button from being tapped during the save process, in the UI, using the Wait class, and that's usually better than a time-based debouncing.

It was just an example with button.
I could move a slider back and forth and save after 2 seconds when I stopped. I cannot block UI in this case.

**Comment by marcglasberg** [comment]
Yes, as I said, the store does that with the persistor. I will see if I can expose this functionality for general use.

**Comment by jans-y** [comment]
For that purpose, I am using this library:

https://pub.dev/packages/easy_debounce

I am wrapping all dispatch() calls that I need to debounce with it.

**Comment by marcglasberg** [comment]
@subzero911 @jans-y 

Hi! Good things come to those who wait. 😬

You can now use the `Debounce` mixin. See here: https://asyncredux.com/flutter/advanced-actions/action-mixins#debounce

There are now many other mixins too, like `Throttle`, `Retry`, `NonReentrant`, `CheckInternet ` and more.

---

## Debounce

Debouncing delays the execution of a function until after a certain period of inactivity.
Each time the debounced function is called, the period of inactivity (or wait time) is reset.

The function will only execute after it stops being called for the duration of the wait time.
Debouncing is useful in situations where you want to ensure that a function is not called too frequently
and only runs after some “quiet time.”

For example, it’s commonly used for handling input validation in text fields,
where you might not want to validate the input every time the user presses a key,
but rather after they've stopped typing for a certain amount of time. For example:

```dart
class SearchText extends AppAction with Debounce {
  
  final String searchTerm;
  SearchText(this.searchTerm);  

  Future<AppState> reduce() async {
      
    var response = await http.get(
      Uri.parse('https://example.com/?q=' + encoded(searchTerm))
    );
        
    return state.copy(searchResult: response.body);
  }
}
```

The `debounce` value is given in milliseconds, and the default is 333 milliseconds (1/3 of a second).
You can override this default:

```dart
class SearchText extends AppAction with Debounce {
    final int debounce = 1000; // Here!
    ...
}
```

### Advanced debounce usage

  The debounce is, by default, based on the action `runtimeType`. This means it will reset the debounce period
  when another action of the same runtimeType was is dispatched within the debounce period. In other words,
  the runtimeType is the "lock". If you want to debounce based on a different lock, you can override
  the `lockBuilder` method. For example, here we debounce two different actions based on the same lock:

  ```dart
  class MyAction1 extends AppAction with Debounce {
     Object? lockBuilder() => 'myLock';
     ...
  }
  
  class MyAction2 extends AppAction with Debounce {
     Object? lockBuilder() => 'myLock';
     ...
  }
  ```

  Another example is to debounce based on some field of the action:

  ```dart
  class MyAction extends AppAction with Debounce {
     final String lock;
     MyAction(this.lock);
     Object? lockBuilder() => lock;
     ...
  }
  ```
